### ISSUE TITLE
Exception contracts for subprograms.

### ISSUE THREAD

**Issue by Richard-Wai** [Ada-Rapporteur-Group/UCI-Prototype#2]
Part of good exception handling often calls for the specific specification of which exceptions can be expected to arise from any given subprogram invocation. By specifying a specific set of possible exceptions, error recovery is much easier to design for users of subprograms for which such a specific set is defined. By using this approach generally and iteratively, it can be quite strait-forward to design highly advanced error recovery mechanisms for a program - particularly for high integrity and/or high reliability systems.

When one is sure of all possible exceptions at a given point, it becomes possible to handle "all" possible exceptional conditions with much greater confidence.

At the present time, the only way to do this is through comments (typically as part of a coding standard). 

In much the same spirit as executable pre/post-conditions, it might be useful to have a formally exception specification contract for subprograms.

For example:

procedure Do_Critical_Action with
  All_Exceptions => Constraint_Error,
  Default_Exception => Constraint_Error;

This construct would specify that Do_Critical_Action may only ever propagate Constraint_Error. Default_Exception, which *must* be specified if All_Exceptions is specified, effectively creates an implicit "exception when others =>" in the exception part of the subprogram body that converts any exception that arises in the subprogram with Default_Exception.

There would be two static (or legality) rules (and no dynamic checks) associated with this:
1. The subprogram cannot explicitly raise anything except one of All_Exceptions
2. Default_Exception must define an Exception that is also a member of All_Exceptions.

These new aspects would be inherited and may not be re-specified.

By using these contracts, programs could know (statically) that any given invocation of such a subprogram will truly never raise any exceptions beyond those specified.

**Comment by sttaft** [comment]
Both C++ and Java have something similar, and alas, the experience with exception specifications has not been great.  Because exceptions such as Storage_Error can arise anywhere, and Constraint_Error is effectively similar given the number of implicit checks, any attempt to say that this is *all* the exceptions that could be raised is pretty much doomed in Ada.  Even in a language like C++ where there are few if any implicit checks, the exception specification feature has been made obsolescent in recent versions.  See for example: https://devblogs.microsoft.com/oldnewthing/20180928-00/?p=99855 and http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2010/n3051.html

**Comment by jprosen** [comment]
I agree with Tuck. Exception contracts do not work well with inheritance. Typical example: if the contract for a stack says it can raise only Stack_Full and Stack_Empty, then you cannot implement a huge stack using a file, since it could raise any IO exception...

**Comment by ARG-Editor** [comment]
Note that a complete proposal for exception contracts is contained in AI12-0017-1. That AI was an unfinished issue from Ada 2022. The ARG resolution of November 10, 2022 including making (or in this case, continuing) an Issue to keep this topic alive for future discussion.

**Comment by ARG-Editor** [comment]
J-P said:

>Exception contracts do not work well with inheritance. Typical example: if the contract for a stack says it can
>raise only Stack_Full and Stack_Empty, then you cannot implement a huge stack using a file, since it could
>raise any IO exception...

One could look at this from the perspective of the user, and note that a stack raising random unexpected exceptions is a problem. Similarly, a stack that only could raise Stack_Full and Stack_Empty is a bug, since there always has to be some exception for an internal failure of some sort. (Cosmic rays happen! And for most compilers, so does out of memory conditions, bounded errors, and the link.)

In general, exception contracts are mainly useful for reusable code. One wants to be told during maintenance if some code change added the possibility of raising an exception that wasn't anticipated during the original design. Otherwise, you are at serious risk of breaking user code. That's the whole reason that we have contracts like types, parameter modes, and postconditions.

The big problem with exception contracts from an Ada perspective is that it isn't reasonable to define them portably. If one assumes that any canconical exception can be raised, then pretty much everything raises Constraint_Error and Program_Error and Storage_Error. And that is both silly (Nat : Natural := 1; possibly raises Constraint_Error canonically, but it is obvious that it does not do so in practice) and pretty much defeats the purpose of exception contracts. And trying to define cases where a compiler should be able to figure out that an exception cannot be raised would be a huge amount of work and also would cause implementation pain by making some implementation models incorrect. I believe that we would have to make it implementation-defined as to when an exception could be raised (the more static analysis a compiler does, the better the job it can make of eliminating exceptions). Whether that would be acceptable to the Ada community is an interesting question in its own right.

                    Randy.,

**Comment by sttaft** [comment]
Java has two classes of exceptions, those which need to be mentioned in exception contracts, and those that don't.  C++ pretty much gave up on exception contracts.

If we think exception contracts would provide a lot of benefit for Ada, we should probably include only a subset of exceptions that require mentioning in contracts, like Java.  We might have an aspect on an exception declaration to indicate whether or not it needs to be mentioned in a contract.

**Comment by clairedross** [comment]
Exception contracts could list exceptions which are expected for a subprogram, and not all that might be raised. Other exceptions could be raised, but they would be more like errors (it is not expected for a user to catch them).

**Comment by CKWG** [comment]
An exceptional case is a program state where the callee cannot fulfil its job
because of some external circumstances which are not under control of the
programmer. Everything else is a programming bug, which also can raise a
(predefined) exception; such a bug, however, has to be fixed and must not occur
in a correct program.

Examples of true exceptional cases:
* A file to be opened does not exist.
* The user has entered an invalid input.
* A sensor sends an invalid value.
* Another program part has failed (tasking).

The caller must be informed about this situation because also he cannot continue
with his job. Programmer-defined exceptions are made for this.

Only these exceptions should be considered for exception contracts.

Examples for bugs:
* Dereferencing a null pointer
* Access to a component that does not exist for the discriminant’s value

Bugs raise the predefined exceptions Constraint_Error, Program_Error,
Storage_Error, Tasking_Error; they must not be used for true exceptional cases
as above.

Since they may occur anywhere, they cannot be part of exception contracts.

**Comment by clairedross** [comment]
Hi all, I don't know if somebody pointed that out already, but SPARK already has exceptional contracts. There are described in the RM https://docs.adacore.com/live/wave/spark2014/html/spark2014_rm/subprograms.html#exceptional-cases and in the user guide https://docs.adacore.com/live/wave/spark2014/html/spark2014_ug/en/source/subprogram_contracts.html#exceptional-contracts. It can be used both to list exceptions that can be propagated and to supply a relevant postcondition. We choose a postcondition as opposed to a precondition because postconditions can model preconditions too, you can get the rationale here https://github.com/clairedross/ada-spark-rfcs/blob/spark-exceptional-contracts/considered/rfc-spark-exceptional-contracts.md. For now, there is no associated runtime semantics though.

**Comment by kevlar700** [comment]
> https://docs.adacore.com/live/wave/spark2014/html/spark2014_ug/en/source/subprogram_contracts.html#exceptional-contracts.

You have to use a package global variable for Index? That seems wrong.

**Comment by sttaft** [comment]
> Hi all, I don't know if somebody pointed that out already, but SPARK already has exceptional contracts.

Thanks, Claire!  Clearly we should consider the SPARK construct as a useful starting point if we decide to pursue this ...

**Comment by sttaft** [comment]
> > https://docs.adacore.com/live/wave/spark2014/html/spark2014_ug/en/source/subprogram_contracts.html#exceptional-contracts.
> 
> You have to use a package global variable for Index? That seems wrong.

This is just an example.  The SPARK construct does not require the use of global variables.
